ksql-api/src/main/java/io/confluent/ksql/api/server/AcksSubscriber.java

```diff
@@ -54,22 +55,41 @@ protected void afterSubscribe(final Subscription subscription) {
   }
 
   @Override
-  public void handleValue(final JsonObject value) {
+  public void handleValue(final InsertResult result) {
     checkContext();
-    insertsStreamResponseWriter.writeInsertResponse();
-    acksSent++;
-    if (insertsSent != null && insertsSent == acksSent) {
-      close();
-    } else if (response.writeQueueFull()) {
-      if (!drainHandlerSet) {
-        response.drainHandler(v -> {
-          drainHandlerSet = false;
-          checkMakeRequest();
-        });
-        drainHandlerSet = true;
+    if (responseEnded) {
+      return;
+    }
+    if (result.succeeded()) {
+      insertsStreamResponseWriter.writeInsertResponse(new InsertAck(result.sequenceNumber()));
+      acksSent++;
+      if (insertsSent != null && insertsSent == acksSent) {
+        close();
+      } else if (response.writeQueueFull()) {
+        if (!drainHandlerSet) {
+          response.drainHandler(v -> {
+            drainHandlerSet = false;
+            checkMakeRequest();
+          });
+          drainHandlerSet = true;
+        }
+      } else {
+        checkMakeRequest();
       }
     } else {
-      checkMakeRequest();
+      log.error("Error in processing inserts", result.exception());
+      final InsertError insertError;
+      final Exception exception = result.exception();
+      if (exception instanceof KsqlInsertsException) {
+        insertError = new InsertError(result.sequenceNumber(),
+            ((KsqlInsertsException) exception).getErrorCode(),
+            exception.getMessage());
+      } else {
+        insertError = new InsertError(result.sequenceNumber(), ERROR_CODE_INTERNAL_ERROR,
+            "Error in processing inserts");
```

Comment:
Not a very helpful error message. Not sure what a user can do with such a message. Can we not provide more details? e.g. the exception message?  Or are we intentionally trying to avoid returning potentially embarrassing error messages? e.g. `null` in a NPE?

Reply:
Right. Any kind of user error (e.g. missing key, wrong types etc) due to bad input should never result in this error. This should only occur if there is some kind of internal error which could be due to a programming error (e.g. NPE) or maybe some environment issue on the server (problems accessing network or filesystem), in which case the real exception will be logged and the user will be sent a simple error to avoid exposing internals on a public API.